src/all/namicomi/src/eu/kanade/tachiyomi/extension/all/namicomi/Namicomi.kt

```diff
@@ -0,0 +1,386 @@
+
+            hasNextPage = newChapterList.hasNextPage
+        }
+
+        // If there are no chapters, don't attempt to check gating
+        if (chapterListResults.isEmpty()) {
+            return emptyList()
+        }
+
+        val gatingCheckRequest = accessibleChapterListRequest(chapterListResults.map { it.id })
+        val gatingCheckResponse = client.newCall(gatingCheckRequest).execute()
+        val accessibleChapterMap = gatingCheckResponse.parseAs<EntityAccessMapDto>()
+            .data?.attributes?.map ?: emptyMap()
+
+        return chapterListResults.mapNotNull {
+            val isAccessible = accessibleChapterMap.getOrElse(it.id) { true }
+            when {
+                // Chapter can be viewed
+                isAccessible -> helper.createChapter(it)
+                // Chapter cannot be viewed and user wants to see locked chapters
+                preferences.showLockedChapters -> {
+                    helper.createChapter(it).apply {
+                        name = helper.intl.format("chapter_locked", name)
+                    }
+                }
+                // Ignore locked chapters otherwise
+                else -> null
+            }
+        }
+    }
+
+    override fun getChapterUrl(chapter: SChapter): String = baseUrl + chapter.url
+
+    override fun pageListRequest(chapter: SChapter): Request {
+        val chapterId = chapter.url.substringAfter("/chapter/")
+        val url = "${NamicomiConstants.apiUrl}/images/chapter/$chapterId?newQualities=true"
+        return GET(url, headers, CacheControl.FORCE_NETWORK)
+    }
+
+    override fun pageListParse(response: Response): List<Page> {
+        val chapterId = response.request.url.pathSegments.last()
+        val pageListDataDto = response.parseAs<PageListDto>().data ?: return emptyList()
+
+        val hash = pageListDataDto.hash
+        val prefix = "${pageListDataDto.baseUrl}/chapter/$chapterId/$hash"
+
+        val urls = if (preferences.useDataSaver) {
+            pageListDataDto.low.map { prefix + "/low/${it.filename}" }
+        } else {
+            pageListDataDto.source.map { prefix + "/source/${it.filename}" }
+        }
+
+        return urls.mapIndexed { index, url ->
+            Page(index, url, url)
+        }
+    }
+
+    override fun imageUrlParse(response: Response): String = ""
+
+    override fun setupPreferenceScreen(screen: PreferenceScreen) {
+        val coverQualityPref = ListPreference(screen.context).apply {
+            key = NamicomiConstants.getCoverQualityPreferenceKey(extLang)
+            title = helper.intl["cover_quality"]
+            entries = NamicomiConstants.getCoverQualityPreferenceEntries(helper.intl)
+            entryValues = NamicomiConstants.getCoverQualityPreferenceEntryValues()
+            setDefaultValue(NamicomiConstants.getCoverQualityPreferenceDefaultValue())
+            summary = "%s"
+
+            setOnPreferenceChangeListener { _, newValue ->
```

Comment:
`setOnPreferenceChangeListener` shouldn't be necessary as app saves the preference itself